ampd/src/queue/queued_broadcaster.rs

```diff
@@ -152,6 +122,52 @@ where
     }
 }
 
+async fn handle_msg<T>(
+    batch_gas_limit: u64,
+    queue: &mut MsgQueue,
+    broadcaster: &mut T,
+    interval: &mut time::Interval,
+    msg_and_res_chan: MsgAndResChan,
+) -> Result
+where
+    T: Broadcaster,
+{
+    let (msg, tx) = msg_and_res_chan;
+
+    match broadcaster.estimate_fee(vec![msg.clone()]).await {
+        Ok(fee) => {
+            tx.send(Ok(())).map_err(|_| Report::new(Error::Client))?;
+
+            if fee.gas_limit.saturating_add(queue.gas_cost()) >= batch_gas_limit {
+                warn!(
+                    queue_size = queue.len(),
+                    queue_gas_cost = queue.gas_cost(),
+                    "exceeded batch gas limit. gas limit can be adjusted in ampd config"
+                );
+                broadcast_all(queue, broadcaster).await?;
+                interval.reset();
+            }
+
+            let message_type = msg.type_url.clone();
+            queue
+                .push(msg, fee.gas_limit)
+                .change_context(Error::Queue)?;
+            info!(
+                message_type,
+                queue_size = queue.len(),
+                queue_gas_cost = queue.gas_cost(),
+                "pushed a new message into the queue"
+            );
```

Comment:
I'd keep the rest in the `run` function and only extract the part that deals with the queue and broadcast